### PR TITLE
Upgrade to support Sniffer 3.x with tests for all sniffs

### PR DESCRIPTION
**Pull request by ewbarnard**
* The full package should now be compatible with squizlabs/php_codesniffer "^3.0". 
* I added unit tests for all Sniffs in this package, using the PHP_CodeSniffer test runner.
* I added information to docs/README.md on the testing structure.
* The sniffer (cs-check and cs-fix) run clean. cs-check and cs-clean need to ignore the entire unit tests directory, since the test cases (fixtures) are intermingled with the unit tests. That's how the PHP_CodeSniffer test runner expects things.
* All unit tests pass (run green).
* I have not tried to use this package as a dependency of anything else.

### CODE REVIEW

**Comment by dereuromark** [comment]
Wow, impressive
So the https://github.com/php-fig-rectified/psr2r-sniffer/pull/7 wasnt of much use then as branch to start off with?

File: PSR2R/Sniffs/Classes/BraceOnSameLineSniff.php [hunk 1]
```diff
@@ -2,32 +2,22 @@
...
 
-use PHP_CodeSniffer_File;
+use PHP_CodeSniffer\Files\File;
+use PHP_CodeSniffer\Sniffs\Sniff;
 
 /**
  * Makes sure opening braces are on the same line for class, interface and trait.
  *
- * @author Mark Scherer
+ * @author  Mark Scherer
```
Comment (dereuromark) [inline PSR2R/Sniffs/Classes/BraceOnSameLineSniff.php]: Must be an accident, as we dont use inline alignment.

File: PSR2R/Sniffs/Classes/InterfaceNameSniff.php [hunk 2]
```diff
@@ -11,18 +12,12 @@ class InterfaceNameSniff extends AbstractSniff {
 
 	/**
 	 * @inheritDoc
+	 * @return void
```
Comment (dereuromark) [inline PSR2R/Sniffs/Classes/InterfaceNameSniff.php]: With inheritDoc the return statement shouldnt be necessary, right?

File: docs/generate.php [hunk 3]
```diff
@@ -3,7 +3,7 @@
 
 exec('vendor/bin/phpcs --standard=PSR2R/ruleset.xml -e', $output, $ret);
 if ($ret !== 0) {
-	die('Invalid execution. Run from ROOT after composer install etc as `php docs/generate.php`.');
+	exit('Invalid execution. Run from ROOT after composer install etc as `php docs/generate.php`.');
```
Comment (dereuromark) [inline docs/generate.php]: don't we use die() as primary one of those two?

File: PSR2R/Sniffs/Classes/BraceOnSameLineSniff.php (hunk 1 above)
Comment (ewbarnard) [inline PSR2R/Sniffs/Classes/BraceOnSameLineSniff.php]: PhpStorm "Reformat Code" does that. I need to see if there's a setting I can adjust. There are likely many places that way. I agree, it goes against the PSR2R standard/recommendations.

File: PSR2R/Sniffs/Classes/InterfaceNameSniff.php (hunk 2 above)
Comment (ewbarnard) [inline PSR2R/Sniffs/Classes/InterfaceNameSniff.php]: I did not think so, but one of the PSR2R sniffs complains without it. The unit tests do not show that error, so it must be one of the other rules mentioned in ruleset.xml. So far I've not figured out which rule is the problem.

Comment (dereuromark) [inline PSR2R/Sniffs/Classes/InterfaceNameSniff.php]: ok

File: docs/generate.php (hunk 3 above)
Comment (ewbarnard) [inline docs/generate.php]: No, exit is primary per PHP/RemoveFunctionAliasSniff.php line 33.

**Comment by ewbarnard** [comment]
Forking "master" rather than "develop" was my mistake. I assumed the more up-to-date branch would be the better starting point. This is my first OSS PR in many years... I don't yet know the significance of travis-ci failing, for example. But I do want to be helpful! I'm expecting to rework the code to clean up anything you notice.

**Comment by ewbarnard** [comment]
Oh, okay, I see why the build failed. It's sniffing tests, which I have as being ignored. I need to figure out how to have "composer cs-check" look at tests (*.php) but ignore fixures (*.inc, *.inc.fixed) which are in the same directories so that I can get that to run cleanly. Then I can update .travis.yml to match cs-check.

Also, I don't see a semantic version defined anywhere. Whatever the version, we should probably bump the major number. It's a jump to sniffer 3.x.

**Comment by dereuromark** [comment]
Yes we can make a Major bumb

File: PSR2R/Sniffs/Classes/InterfaceNameSniff.php (hunk 2 above)
Comment (ewbarnard) [inline PSR2R/Sniffs/Classes/InterfaceNameSniff.php]: Looks like it is PSR2R.Commenting.DocBlockReturnTag.Missing. I'll fix the sniff.

**Comment by ewbarnard** [comment]
@dereuromark Mark, all looks clean now and the build passes. Much improved, thanks for the review! The test setup makes it easy to TDD sniff changes. Just add a sample to the fixture (e.g., combine inheritDoc and return tags), watch it fail, and fix the sniff.

**Comment by dereuromark** [comment]
Nice job!
We can now finalize in master until stable enough to tag as 1.0.0 release :)

**Comment by ewbarnard** [comment]
I'm pretty sure this is my first OSS code contribution of this millennium. Thank you! I'll take a look at cakephp-ide-helper with PHP_CodeSniffer 3.x next, which was the point of this exercise.

**Comment by dereuromark** [comment]
Ideally we can wrap the sniffer access in a way that both 2 and 3 version of it are working.
This way people can use it with older and newer sniffer.

**Pull request merged**